src/hive_mind_bridge/src/HiveMindBridgeImpl.cpp

```diff
@@ -2,23 +2,46 @@
 
 HiveMindBridgeImpl::HiveMindBridgeImpl(ITCPServer& tcpServer,
                                        IHiveMindHostSerializer& serializer,
-                                       IHiveMindHostDeserializer& deserializer) :
-    m_tcpServer(tcpServer), m_serializer(serializer), m_deserializer(deserializer) {}
+                                       IHiveMindHostDeserializer& deserializer,
+                                       IMessageHandler& messageHandler,
+                                       IThreadSafeQueue<MessageDTO>& queue) :
+    m_tcpServer(tcpServer),
+    m_serializer(serializer),
+    m_deserializer(deserializer),
+    m_messageHandler(messageHandler),
+    m_inboundQueue(queue) {}
 
 void HiveMindBridgeImpl::spin() {
-    if (m_tcpServer.isClientConnected()) {
-        MessageDTO message;
-        if (m_deserializer.deserializeFromStream(message)) {
+    if (isTCPClientConnected()) {
+        if (!m_inboundQueue.empty()) {
             // Execute the action
-            MessageDTO responseMessage = m_messageHandler.handleMessage(message);
+            MessageHandlerResult result = m_messageHandler.handleMessage(m_inboundQueue.front());
+            m_inboundQueue.pop();
+
+            m_resultQueue.push_back(result);
 
             // Send the ack/nack message
-            m_serializer.serializeToStream(responseMessage);
-        } else {
-            ROS_WARN("The received bytes do not contain a valid message.");
+            m_serializer.serializeToStream(result.getResponse());
+        }
+
+        for (auto result = m_resultQueue.begin(); result != m_resultQueue.end();) {
+            if (result->getCallbackReturnContext().wait_for(std::chrono::seconds(0)) ==
+                std::future_status::ready) {
+                sendReturn(*result);
+
+                m_resultQueue.erase(result);
+            } else {
+                result++;
+            }
         }
     } else {
+        if (m_inboundThread.joinable()) {
```

Comment:
_joinable_ call not required if _join_ is called right after.

Reply:
Yes it is. When the `BridgeImpl` object is created, is creates a thread using std::thread's default constructor. This only creates a sort of "placeholder" thread object, that does not really run. At that moment, it is not joinable.

Bottom line : when the program launches, the TCP client is not connected, and the thread has not started, so we don't want to join at that time (and it would be impossible since the thread is not joinable). We only join after the TCP connection has died and it is time to start a new one and as well as a new thread.